File: tables/scripts/stage1_ocr_hint_table.R
#!/usr/bin/env Rscript
#
# Build the Stage 1 OCR-hint ablation LaTeX table, comparing each dictionary's best model/alphabet combo with vs. without an OCR-system hint in the prompt.
#
# Usage:
#   Rscript evaluations/scripts/stage1_ocr_hint_table.R
#
# Output:
#   evaluations/tables/stage1_ocr_hint_per_dictionary.tex

suppressWarnings(suppressMessages({
  script_path <- commandArgs(trailingOnly = FALSE)
  script_arg <- sub("^--file=", "", script_path[grep("^--file=", script_path)])
  repo_root <- if (length(script_arg)) normalizePath(file.path(dirname(script_arg), "..", "..")) else getwd()
}))

hint_csv <- file.path(repo_root, "evaluations", "stage1_flat_ocrhint_best_eval", "stage1_flat_eval_summary.csv")
baseline_csv <- file.path(repo_root, "evaluations", "stage1_flat_per_lang_script_eval", "stage1_flat_eval_summary.csv")
output_dir <- file.path(repo_root, "tables", "tables")
output_path <- file.path(output_dir, "stage1_ocr_hint_per_dictionary.tex")

if (!file.exists(hint_csv)) {
  stop(sprintf(
    "OCR-hint summary CSV not found: %s\nGenerate it first with:\n  uv run mudidi benchmark evaluate stage1 --stage1-output-subdir stage-1-ocr --all-experiments --dataset-dir dataset/mudidi/dictionaries --pred-root outputs/benchmark/stage-1 -o evaluations/stage1_flat_ocrhint_best_eval",
    hint_csv
  ))
}
if (!file.exists(baseline_csv)) {
  stop(sprintf("Baseline (no-hint) summary CSV not found: %s", baseline_csv))
}

hint_results <- read.csv(hint_csv, stringsAsFactors = FALSE)
baseline_results <- read.csv(baseline_csv, stringsAsFactors = FALSE)

metric_cols <- c("TextEdit", "GCER", "WER", "typography_f1", "ReadOrderEdit")
higher_is_better <- c(TextEdit = FALSE, GCER = FALSE, WER = FALSE,
                       typography_f1 = TRUE, ReadOrderEdit = FALSE)

# --- model key (from experiment name prefix) -> short display name used in this table ---
model_display <- c(
  claudeopus47 = "Claude-Opus",
  gemini31pro = "Gemini-Pro",
  gemini3flash = "Gemini-Flash",
  gpt55 = "GPT-5.5"
)

model_key_from_experiment <- function(experiment) {
  # e.g. "claudeopus47_flat_alpha_ocrhint" -> "claudeopus47"
  sub("_flat_.*$", "", experiment)
}

hint_results$model_key <- model_key_from_experiment(hint_results$experiment)
unknown_models <- setdiff(hint_results$model_key, names(model_display))
if (length(unknown_models)) {
  stop(sprintf("Unrecognized model key(s) in OCR-hint experiment names: %s",
               paste(unknown_models, collapse = ", ")))
}

# alphabet is read from run_config.json by the evaluator (authoritative) --
# don't infer it from the "_alpha"/"_noalpha" substring in the experiment name.
hint_results$alphabet_flag <- tolower(trimws(hint_results$alphabet)) == "true"
hint_results$baseline_experiment <- sprintf(
  "%s_flat_%s", hint_results$model_key,
  ifelse(hint_results$alphabet_flag, "alpha", "noalpha")
)

# --- join hint rows to their matching no-hint baseline row (same dictionary + baseline experiment) ---
baseline_keyed <- baseline_results[, c("experiment", "language", metric_cols)]
names(baseline_keyed)[names(baseline_keyed) %in% metric_cols] <- paste0("base_", metric_cols)

merged <- merge(
  hint_results,
  baseline_keyed,
  by.x = c("baseline_experiment", "language"),
  by.y = c("experiment", "language"),
  all.x = TRUE
)
missing_baseline <- merged$language[is.na(merged$base_TextEdit)]
if (length(missing_baseline)) {
  stop(sprintf(
    "No matching no-hint baseline found for: %s",
    paste(sprintf("%s (wanted %s)", missing_baseline,
                   merged$baseline_experiment[is.na(merged$base_TextEdit)]),
          collapse = "; ")
  ))
}
merged <- merged[order(merged$language), ]

fmt <- function(x) sprintf("%.3f", x)

# Display-only correction: the dataset folder / CSVs spell this dictionary
# with an underscore ("Kurdish_Turkish"); it should render with a hyphen.
# Data lookups still key on the original underscore spelling.
display_dictionary_name <- function(x) sub("Kurdish_Turkish", "Kurdish-Turkish", x, fixed = TRUE)

# Bold whichever side of the pair is better, decided on full-precision values;
# exact ties (as seen in practice, e.g. WER 0.027 vs 0.027) are left unbolded.
pair_cells <- function(base_val, hint_val, higher_better) {
  if (base_val == hint_val) {
    return(c(fmt(base_val), fmt(hint_val)))
  }
  base_wins <- if (higher_better) base_val > hint_val else base_val < hint_val
  c(
    if (base_wins) sprintf("\\textbf{%s}", fmt(base_val)) else fmt(base_val),
    if (!base_wins) sprintf("\\textbf{%s}", fmt(hint_val)) else fmt(hint_val)
  )
}

row_line <- function(i) {
  r <- merged[i, ]
  checkmark <- if (isTRUE(r$alphabet_flag)) "\\cmark" else ""
  cells <- unlist(lapply(metric_cols, function(col) {
    pair_cells(r[[paste0("base_", col)]], r[[col]], higher_is_better[[col]])
  }))
  sprintf(
    "%s & %s & %s & %s \\\\",
    display_dictionary_name(r$language), model_display[[r$model_key]], checkmark,
    paste(cells, collapse = " & ")
  )
}

data_lines <- vapply(seq_len(nrow(merged)), row_line, character(1))

# --- Mean row: unweighted mean across all 30 dictionaries; bold whichever
# side (no-hint vs. hint) wins per metric, same pair_cells logic as the
# per-dictionary rows above.
mean_cells <- unlist(lapply(metric_cols, function(col) {
  pair_cells(
    mean(merged[[paste0("base_", col)]]),
    mean(merged[[col]]),
    higher_is_better[[col]]
  )
}))
mean_line <- sprintf("\\textit{Mean} &  &  & %s \\\\", paste(mean_cells, collapse = " & "))

lines <- c(
  "% Auto-generated by evaluations/scripts/stage1_ocr_hint_table.R -- do not edit by hand.",
  "\\section{Stage-1 OCR Assisted Prompting Results}",
  "\\label{sec:app-ocr-asst}",
  "\\begin{table*}[!h]",
  "\\centering",
  "\\scriptsize",
  "\\setlength{\\tabcolsep}{3pt}",
  "\\renewcommand{\\arraystretch}{1.08}",
  "\\begin{adjustbox}{max width=\\textwidth}",
  "\\begin{tabular}{llc cc cc cc cc cc}",
  "\\toprule",
  " & & & \\multicolumn{2}{c}{\\textbf{Edit}} & \\multicolumn{2}{c}{\\textbf{GCER}} & \\multicolumn{2}{c}{\\textbf{WER}} & \\multicolumn{2}{c}{\\textbf{Mrk. F1}} & \\multicolumn{2}{c}{\\textbf{Order}} \\\\",
  "\\cmidrule(lr){4-5}\\cmidrule(lr){6-7}\\cmidrule(lr){8-9}\\cmidrule(lr){10-11}\\cmidrule(lr){12-13}",
  "\\textbf{Dictionary} & \\textbf{Model} & \\textbf{A} & -- & +hint & -- & +hint & -- & +hint & -- & +hint & -- & +hint \\\\",
  "\\midrule",
  data_lines,
  "\\midrule",
  mean_line,
  "\\bottomrule",
  "\\end{tabular}",
  "\\end{adjustbox}",
  "\\caption{Per-dictionary breakdown of the Stage 1 OCR-hint ablation summarised in Table~\\ref{tab:stage1-aggregate-ocr-hint}. Each metric is shown as a paired (without hint, with hint) value. \\textit{Best score per pair is bolded; lower is better except for Markup F1.}}",
  "\\label{tab:stage1-ocr-hint-per-language}",
  "\\end{table*}"
)

dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
writeLines(lines, output_path)

cat(sprintf("Wrote %s (%d dictionaries)\n", output_path, nrow(merged)))

# ============================================================
# Table 2: pooled OCR-hint ablation -- same "best config per dictionary"
# data as above, averaged over all 30 dictionaries into a single with/without
# comparison row, plus a delta row.
# ============================================================

pooled_output_path <- file.path(output_dir, "stage1_ocr_hint_summary.tex")
fmt2 <- function(x) sprintf("%.2f", x)

mean_no_hint <- vapply(metric_cols, function(col) mean(merged[[paste0("base_", col)]]), numeric(1))
mean_hint <- vapply(metric_cols, function(col) mean(merged[[col]]), numeric(1))

# Best-score bolding per metric, decided on full-precision means; exact ties left unbolded.
pooled_pair_cells <- function(no_hint_val, hint_val, higher_better) {
  if (no_hint_val == hint_val) {
    return(c(fmt2(no_hint_val), fmt2(hint_val)))
  }
  no_hint_wins <- if (higher_better) no_hint_val > hint_val else no_hint_val < hint_val
  c(
    if (no_hint_wins) sprintf("\\textbf{%s}", fmt2(no_hint_val)) else fmt2(no_hint_val),
    if (!no_hint_wins) sprintf("\\textbf{%s}", fmt2(hint_val)) else fmt2(hint_val)
  )
}

no_hint_row_cells <- character(0)
hint_row_cells <- character(0)
delta_cells <- character(0)
for (col in metric_cols) {
  pair <- pooled_pair_cells(mean_no_hint[[col]], mean_hint[[col]], higher_is_better[[col]])
  no_hint_row_cells <- c(no_hint_row_cells, pair[1])
  hint_row_cells <- c(hint_row_cells, pair[2])
  delta_cells <- c(delta_cells, sprintf("$%+.2f$", mean_hint[[col]] - mean_no_hint[[col]]))
}

pooled_lines <- c(
  "% Auto-generated by evaluations/scripts/stage1_ocr_hint_table.R -- do not edit by hand.",
  "% ============================================================",
  "% Table 2: Pooled OCR-hint ablation",
  "% For each of the 30 dictionaries, take the strongest LLM+alphabet",
  "% configuration identified in Table 1 and compare with vs. without",
  "% an auxiliary OCR transcript supplied at inference.",
  "% ============================================================",
  "\\begin{table}[!h]",
  "\\centering",
  "\\small",
  "\\setlength{\\tabcolsep}{3pt}",
  "%\\renewcommand{\\arraystretch}{1.1}",
  "\\caption{Stage 1 OCR-hint ablation study, averaged over 30 dictionaries. For each dictionary, we hold fixed the strongest LLM and alphabet configuration from Table~\\ref{tab:stage1-aggregate-alphabet} and compare transcription with vs.\\ without a preliminary OCR transcript supplied to the model. \\textit{Best score per metric is bolded.}}",
  "\\label{tab:stage1-aggregate-ocr-hint}",
  "%\\begin{adjustbox}{max width=\\columnwidth}",
  "\\begin{tabular}{lcrrrrr}",
  "\\toprule",
  "\\textbf{Configuration} & \\textbf{OCR hint} & \\textbf{Edit} & \\textbf{GCER} & \\textbf{WER} & \\textbf{Mrk. F1} & \\textbf{Order} \\\\",
  "\\midrule",
  sprintf("Best LLM + alphabet per language &  & %s \\\\", paste(no_hint_row_cells, collapse = " & ")),
  sprintf("Best LLM + alphabet per language & \\cmark & %s \\\\", paste(hint_row_cells, collapse = " & ")),
  "\\midrule",
  sprintf("$\\Delta$ (with hint $-$ without)  &  & %s \\\\", paste(delta_cells, collapse = " & ")),
  "\\bottomrule",
  "\\end{tabular}",
  "%\\end{adjustbox}",
  "\\end{table}"
)

writeLines(pooled_lines, pooled_output_path)

cat(sprintf("Wrote %s (pooled over %d dictionaries)\n", pooled_output_path, nrow(merged)))

File: tables/scripts/stage1_ocr_hint_table.py
#!/usr/bin/env python3
"""Generate fixed Gemini 3.1 Pro + alphabet OCR-hint ablation tables."""

from __future__ import annotations

from collections.abc import Sequence
from pathlib import Path
from statistics import fmean

from table_utils import (
    TableDataError,
    index_unique,
    lines_to_text,
    parse_bool,
    parse_float,
    parse_paths,
    read_delimited,
    write_text_atomic,
)

BASELINE_EXPERIMENT = "gemini31pro_flat_alpha"
HINT_EXPERIMENT = "gemini31pro_flat_alpha_ocr"
METRICS = ("TextEdit", "GCER", "WER", "typography_f1", "ReadOrderEdit")
HIGHER_IS_BETTER = {metric: metric == "typography_f1" for metric in METRICS}
REQUIRED_COLUMNS = ("experiment", "language", "alphabet", "ocr-hint", *METRICS)


def _load_experiment(path: Path, experiment: str, *, expect_hint: bool) -> dict[str, dict[str, object]]:
    raw_rows = read_delimited(path, required_columns=REQUIRED_COLUMNS)
    selected = [row for row in raw_rows if row["experiment"] == experiment]
    unexpected = sorted({row["experiment"] for row in raw_rows if row["experiment"] != experiment})
    if unexpected:
        raise TableDataError(f"{path}: unexpected experiment(s): {', '.join(unexpected)}")
    keyed = index_unique(selected, ("language",), source=path)
    if len(keyed) != 30:
        raise TableDataError(f"{path}: {experiment} has {len(keyed)} dictionaries, expected 30")

    parsed: dict[str, dict[str, object]] = {}
    for row_number, row in enumerate(selected, start=2):
        alphabet = parse_bool(row["alphabet"], source=path, field="alphabet", row_number=row_number)
        hint = parse_bool(row["ocr-hint"], source=path, field="ocr-hint", row_number=row_number)
        if not alphabet:
            raise TableDataError(f"{path}: row {row_number}: fixed protocol requires alphabet=true")
        if hint != expect_hint:
            raise TableDataError(
                f"{path}: row {row_number}: expected ocr-hint={str(expect_hint).lower()}"
            )
        parsed[row["language"]] = {
            metric: parse_float(row[metric], source=path, field=metric, row_number=row_number)
            for metric in METRICS
        }
    return parsed


def _paired_rows(repo_root: Path) -> list[tuple[str, dict[str, object], dict[str, object]]]:
    evaluation_root = repo_root / "evaluations" / "stage1_flat_per_lang_script_eval"
    baseline_source = evaluation_root / "stage1_flat_eval_summary.csv"
    hint_source = evaluation_root / "stage1_flat_eval_ocr_hint_summary.csv"

    all_baseline_rows = read_delimited(baseline_source, required_columns=REQUIRED_COLUMNS)
    baseline_only = [row for row in all_baseline_rows if row["experiment"] == BASELINE_EXPERIMENT]
    if len(baseline_only) != 30:
        raise TableDataError(
            f"{baseline_source}: {BASELINE_EXPERIMENT} has {len(baseline_only)} dictionaries, expected 30"
        )
    # Parse the selected baseline rows directly; the canonical summary intentionally
    # contains other Stage 1 experiments.
    index_unique(baseline_only, ("language",), source=baseline_source)
    baseline: dict[str, dict[str, object]] = {}
    for row_number, row in enumerate(baseline_only, start=2):
        alphabet = parse_bool(row["alphabet"], source=baseline_source, field="alphabet", row_number=row_number)
        hint = parse_bool(row["ocr-hint"], source=baseline_source, field="ocr-hint", row_number=row_number)
        if not alphabet or hint:
            raise TableDataError(
                f"{baseline_source}: {row['language']}: fixed baseline requires alphabet=true and ocr-hint=false"
            )
        baseline[row["language"]] = {
            metric: parse_float(
                row[metric], source=baseline_source, field=metric, row_number=row_number
            )
            for metric in METRICS
        }

    hints = _load_experiment(hint_source, HINT_EXPERIMENT, expect_hint=True)
    if set(baseline) != set(hints):
        raise TableDataError(
            "OCR-hint dictionary set differs from the fixed baseline; "
            f"missing={sorted(set(baseline) - set(hints))}, "
            f"extra={sorted(set(hints) - set(baseline))}"
        )
    return [(language, baseline[language], hints[language]) for language in sorted(baseline)]


def _pair_cells(baseline: float, hint: float, *, higher_is_better: bool, digits: int) -> tuple[str, str]:
    baseline_text = f"{baseline:.{digits}f}"
    hint_text = f"{hint:.{digits}f}"
    if baseline == hint:
        return baseline_text, hint_text
    baseline_wins = baseline > hint if higher_is_better else baseline < hint
    return (
        rf"\textbf{{{baseline_text}}}" if baseline_wins else baseline_text,
        rf"\textbf{{{hint_text}}}" if not baseline_wins else hint_text,
    )


def render_tables(repo_root: Path) -> tuple[str, str]:
    """Render the per-dictionary and aggregate fixed-protocol tables."""
    rows = _paired_rows(repo_root)
    data_lines: list[str] = []
    for language, baseline, hint in rows:
        cells: list[str] = []
        for metric in METRICS:
            cells.extend(
                _pair_cells(
                    float(baseline[metric]),
                    float(hint[metric]),
                    higher_is_better=HIGHER_IS_BETTER[metric],
                    digits=3,
                )
            )
        display_language = language.replace("Kurdish_Turkish", "Kurdish-Turkish")
        data_lines.append(
            f"{display_language} & Gemini-Pro & \\cmark & {' & '.join(cells)} \\\\"
        )

    means = {
        "baseline": {metric: fmean(float(baseline[metric]) for _, baseline, _ in rows) for metric in METRICS},
        "hint": {metric: fmean(float(hint[metric]) for _, _, hint in rows) for metric in METRICS},
    }
    mean_cells: list[str] = []
    baseline_cells: list[str] = []
    hint_cells: list[str] = []
    delta_cells: list[str] = []
    for metric in METRICS:
        baseline_value = means["baseline"][metric]
        hint_value = means["hint"][metric]
        pair = _pair_cells(
            baseline_value,
            hint_value,
            higher_is_better=HIGHER_IS_BETTER[metric],
            digits=3,
        )
        mean_cells.extend(pair)
        baseline_cells.append(pair[0])
        hint_cells.append(pair[1])
        delta_cells.append(f"${hint_value - baseline_value:+.3f}$")

    per_dictionary_lines = [
        "% Auto-generated by tables/scripts/stage1_ocr_hint_table.py -- do not edit by hand.",
        r"\section{Stage-1 OCR Assisted Prompting Results}",
        r"\label{sec:app-ocr-asst}",
        r"\begin{table*}[!h]",
        r"\centering",
        r"\scriptsize",
        r"\setlength{\tabcolsep}{3pt}",
        r"\renewcommand{\arraystretch}{1.08}",
        r"\begin{adjustbox}{max width=\textwidth}",
        r"\begin{tabular}{llc cc cc cc cc cc}",
        r"\toprule",
        " & & & \\multicolumn{2}{c}{\\textbf{Edit}} & \\multicolumn{2}{c}{\\textbf{GCER}} & \\multicolumn{2}{c}{\\textbf{WER}} & \\multicolumn{2}{c}{\\textbf{Mrk. F1}} & \\multicolumn{2}{c}{\\textbf{Order}} \\\\",
        r"\cmidrule(lr){4-5}\cmidrule(lr){6-7}\cmidrule(lr){8-9}\cmidrule(lr){10-11}\cmidrule(lr){12-13}",
        "\\textbf{Dictionary} & \\textbf{Model} & \\textbf{A} & -- & +hint & -- & +hint & -- & +hint & -- & +hint & -- & +hint \\\\",
        r"\midrule",
        *data_lines,
        r"\midrule",
        f"\\textit{{Mean}} &  &  & {' & '.join(mean_cells)} \\\\",
        r"\bottomrule",
        r"\end{tabular}",
        r"\end{adjustbox}",
        r"\caption{Per-dictionary breakdown of the Stage 1 OCR-hint ablation summarised in Table~\ref{tab:stage1-aggregate-ocr-hint}. Gemini~3.1 Pro receives the source-language alphabet in both conditions; each metric is shown as a paired (without hint, with hint) value. \textit{Best score per pair is bolded; lower is better except for Markup F1.}}",
        r"\label{tab:stage1-ocr-hint-per-language}",
        r"\end{table*}",
    ]
    summary_lines = [
        "% Auto-generated by tables/scripts/stage1_ocr_hint_table.py -- do not edit by hand.",
        "% Fixed Gemini 3.1 Pro + alphabet OCR-hint ablation, averaged over 30 dictionaries.",
        r"\begin{table}[!h]",
        r"\centering",
        r"\small",
        r"\setlength{\tabcolsep}{3pt}",
        r"\caption{Stage 1 OCR-hint ablation study, averaged over 30 dictionaries. We hold Gemini~3.1 Pro with alphabet input fixed and compare transcription with vs.\ without a preliminary OCR transcript supplied to the model. \textit{Best score per metric is bolded.}}",
        r"\label{tab:stage1-aggregate-ocr-hint}",
        r"\begin{tabular}{lcrrrrr}",
        r"\toprule",
        "\\textbf{Configuration} & \\textbf{OCR hint} & \\textbf{Edit} & \\textbf{GCER} & \\textbf{WER} & \\textbf{Mrk. F1} & \\textbf{Order} \\\\",
        r"\midrule",
        f"Gemini 3.1 Pro + alphabet &  & {' & '.join(baseline_cells)} \\\\",
        f"Gemini 3.1 Pro + alphabet & \\cmark & {' & '.join(hint_cells)} \\\\",
        r"\midrule",
        f"$\\Delta$ (with hint $-$ without) &  & {' & '.join(delta_cells)} \\\\",
        r"\bottomrule",
        r"\end{tabular}",
        r"\end{table}",
    ]
    return lines_to_text(per_dictionary_lines), lines_to_text(summary_lines)


def main(argv: Sequence[str] | None = None) -> None:
    """Generate both fixed-protocol OCR-hint tables."""
    repo_root, output_dir = parse_paths(__file__, __doc__ or "", argv)
    per_dictionary, summary = render_tables(repo_root)
    per_dictionary_path = output_dir / "stage1_ocr_hint_per_dictionary.tex"
    summary_path = output_dir / "stage1_ocr_hint_summary.tex"
    write_text_atomic(per_dictionary_path, per_dictionary)
    write_text_atomic(summary_path, summary)
    print(f"Wrote {per_dictionary_path} (30 dictionaries)")
    print(f"Wrote {summary_path} (fixed Gemini 3.1 Pro + alphabet)")


if __name__ == "__main__":
    main()
